### PR TITLE
[FIX] `load_confounds`: Raise an exception when all volumes are scrubbed

### LINKED ISSUE #4557: [BUG] `load_confounds` returns all nans when `sample_mask` is empty & `demean = True`
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Operating system

- [ ] Linux
- [X] Mac
- [ ] Windows

### Operating system version

MacOS Sequoia 15.0

### Python version

- [ ] 3.12
- [X] 3.11
- [ ] 3.10
- [ ] 3.9
- [ ] 3.8

### nilearn version

0.10.4

### Expected behavior

I expected `load_confounds` to inform me if all volumes are outliers / non-stable-state before quit.

### Current behavior & error messages

Instead, even though `sample_mask.size == 0`, confound loading doesn't stop after `_extract_outlier_regressors`. The consequence is that `_demean_confounds` outputs all nans.

Output:

```bash
>>> confounds_sample
     cosine00  cosine01  cosine02  ...  trans_z_derivative1_power2  trans_z_power2  white_matter
0         NaN       NaN       NaN  ...                         NaN             NaN           NaN
1         NaN       NaN       NaN  ...                         NaN             NaN           NaN
2         NaN       NaN       NaN  ...                         NaN             NaN           NaN
3         NaN       NaN       NaN  ...                         NaN             NaN           NaN
4         NaN       NaN       NaN  ...                         NaN             NaN           NaN
..        ...       ...       ...  ...                         ...             ...           ...
378       NaN       NaN       NaN  ...                         NaN             NaN           NaN
379       NaN       NaN       NaN  ...                         NaN             NaN           NaN
380       NaN       NaN       NaN  ...                         NaN             NaN           NaN
381       NaN       NaN       NaN  ...                         NaN             NaN           NaN
382       NaN       NaN       NaN  ...                         NaN             NaN           NaN
```

```bash
>>> sample_mask.size
0
```

### Steps and code to reproduce bug

The output is produced by:

```python
confounds_sample, sample_mask = load_confounds(
ts_path,
strategy=["high_pass", "motion", "wm_csf", "scrub"],
motion="full",
wm_csf="basic",
scrub=5,
fd_threshold=0.2,
)

confounds_sample
sample_mask.size
```

### PR DESCRIPTION
**Pull request by victoris93**
Closes #4557 

* Added a new exception class `AllVolumesRemoved`
* `AllVolumesRemoved` is raised when `sample_mask` is an empty array, i.e., all volumes are either motion outliers or non-steady-state. This way `load_confounds` stops at `extract_outlier_regressors`.
* Grouped custom exception classes in `interfaces/fmriprep/exceptions.py`
* Modified imports accordingly in `interfaces.fmriprep`

### CODE REVIEW

**Comment by Remi-Gau** [comment]
Need to double check but maybe it's better to throw a warning, no?
Imagine looping over subjects and runs,  I'd rather get a warning and be able to decide how to handle this rather than getting an error that stops me in my track.

**Comment by Remi-Gau** [comment]
FYI to fix the linting issues, run :

```bash
pip install tox
tox
```

And then commit the changes.

**Comment by victoris93** [comment]
> it's better to throw a warning, no? Imagine looping over subjects and runs, I'd rather get a warning and be able to decide how to handle this rather than getting an error that stops me in my track.
>
Yes, I agree, in `prepare_output` we're better off with a warning. 
However, I think this exception is warranted [somewhere here](https://github.com/nilearn/nilearn/blob/4f4730163097457cf9ddb5674ffd158ee8fa822e/nilearn/signal.py#L879) because currently `masker.fit_transform` throws an error about `scipy.interpolate.CubicSpline` not giving much information about the inputs on the `nilearn` end:

```python
atlas = datasets.fetch_atlas_schaefer_2018(n_rois=400, yeo_networks=7, resolution_mm=1, base_url= None, resume=True, verbose=1)
bold_tr = 0.8

confounds_sample, sample_mask = load_confounds(
ts_path,
strategy=["high_pass", "motion", "wm_csf", "scrub"],
motion="full",
wm_csf="basic",
scrub=5,
fd_threshold=0.2,
demean = False
)

masker =  NiftiLabelsMasker(labels_img=atlas.maps, standardize="zscore_sample", standardize_confounds="zscore_sample", low_pass=0.08, high_pass=0.009, memory='nilearn_cache', verbose=5, smoothing_fwhm=2, t_r = bold_tr)
parc_ts = masker.fit_transform(ts_path, confounds = confounds_sample, sample_mask = sample_mask)
```

```bash
File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/sklearn/utils/_set_output.py", line 316, in wrapped
    data_to_wrap = f(self, X, *args, **kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/nilearn/maskers/nifti_labels_masker.py", line 620, in fit_transform
    return self.fit(imgs).transform(
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/sklearn/utils/_set_output.py", line 316, in wrapped
    data_to_wrap = f(self, X, *args, **kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/nilearn/maskers/base_masker.py", line 268, in transform
    return self.transform_single_imgs(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/nilearn/maskers/nifti_labels_masker.py", line 724, in transform_single_imgs
    region_signals, (ids, masked_atlas) = self._cache(
                                          ^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/joblib/memory.py", line 577, in __call__
    return self._cached_call(args, kwargs, shelving=False)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/joblib/memory.py", line 532, in _cached_call
    return self._call(call_id, args, kwargs, shelving)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/joblib/memory.py", line 771, in _call
    output = self.func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/nilearn/maskers/base_masker.py", line 139, in _filter_and_extract
    region_signals = cache(
                     ^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/joblib/memory.py", line 312, in __call__
    return self.func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/nilearn/signal.py", line 743, in clean
    signals, confounds, sample_mask = _handle_scrubbed_volumes(
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/nilearn/signal.py", line 836, in _handle_scrubbed_volumes
    signals = _interpolate_volumes(signals, sample_mask, t_r, extrapolate)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/nilearn/signal.py", line 879, in _interpolate_volumes
    cubic_spline_fitter = CubicSpline(
                          ^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/scipy/interpolate/_cubic.py", line 751, in __init__
    x, dx, y, axis, _ = prepare_input(x, y, axis)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/gpfs3/well/margulies/users/cpy397/python/neuro/lib/python3.11/site-packages/scipy/interpolate/_cubic.py", line 55, in prepare_input
    raise ValueError("`x` must contain at least 2 elements.")
ValueError: `x` must contain at least 2 elements.
```

If you think this exception is useful here, would you like it as a separate PR?

File: nilearn/interfaces/fmriprep/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+
+class AllVolumesRemoved(Exception):
+    """
+    Exception raised when all volumes have either motion outliers or are non-steady-state volumes (`sample_mask` is an empty array).
+    """
+
+    def __init__(self):
+        super().__init__(
+            "All volumes are either outliers or non-steady-state. The size of the sample mask is 0."
```
Comment (Remi-Gau) [inline nilearn/interfaces/fmriprep/exceptions.py:24]: I think I would rephrase this part of the message into something:
"All time points in the confounds were marked as outliers. This would lead to all volumes in the time series to be scrubbed."
Just in case the confounds people pass are not just about non steady state.

**Review by Remi-Gau — CHANGES_REQUESTED** [review]

**Comment by Remi-Gau** [comment]
A few more things to do.
- add tests to make sure that new errors and warnings are thrown in the right circumstances
- make sure those errors / warnings are not raised by other tests
- update the changelog and add yourself as Contributor.

**Comment by Remi-Gau** [comment]
@htwangtw pinging you so you follow and review this one as this up your alley I believe and you may correct any sillyness coming from my side.

File: nilearn/interfaces/fmriprep/exceptions.py (hunk 1 above)
Comment (htwangtw) [reply nilearn/interfaces/fmriprep/exceptions.py:24]: I second Remi. The case of all volumes are marked as non steady state is extremely unlikely. 

```
All time points in the confounds were marked as outliers. This would lead to all volumes in the time series to be scrubbed. 
```

**Comment by htwangtw** [comment]
unrelated to this PR, but in the example you have here, if you are already including cosine regressors, you should not include any low/high pass filters parameters.

**Comment by victoris93** [comment]
> if you are already including cosine regressors, you should not include any low/high pass filters parameters.

Huge thanks for bringing this to my attention!

**Comment by Remi-Gau** [comment]
doc building is failing because some warning are being thrown by sphinx: let me check and see where this is coming from.

https://github.com/nilearn/nilearn/actions/runs/11030148676/job/30634251328#step:14:1638

File: doc/changes/latest.rst [hunk 2]
```diff
@@ -43,6 +43,9 @@ Enhancements
 
 Changes
 -------
+- :bdg-dark:`Code` Warn the user :func:`nilearn.interfaces.fmriprep.load_confounds_scrub.extract_outlier_regressors` if ``sample_mask`` is empty (all volumes are scrubbed) (:gh:`4558` by `Victoria Shevchenko`_).
+
+- :bdg-dark:`Code` Throw error in :func:`nilearn.signal._handle_scrubbed_volumes` if ``sample_mask`` is empty (:gh:`4558` by `Victoria Shevchenko`_).
```
Comment (Remi-Gau) [inline doc/changes/latest.rst:48]:
```suggestion
- :bdg-dark:`Code` Throw error if ``sample_mask`` is empty when scrubbing an fMRI time series (:gh:`4558` by `Victoria Shevchenko`_).
```

I suspect the doc build was failing because we were referencing a private function: not naming the function should solve this (also private function can be renamed at any time so this way we do not have to update the changelog if this function is renamed in the future).

File: doc/changes/latest.rst [hunk 3]
```diff
@@ -43,6 +43,9 @@ Enhancements
 
 Changes
 -------
+- :bdg-dark:`Code` Warn the user :func:`nilearn.interfaces.fmriprep.load_confounds_scrub.extract_outlier_regressors` if ``sample_mask`` is empty (all volumes are scrubbed) (:gh:`4558` by `Victoria Shevchenko`_).
```
Comment (Remi-Gau) [inline doc/changes/latest.rst:46]:
```suggestion
- :bdg-dark:`Code` Warn the user when all volumes would be scrubbed when loading fmriprep confounds as this would lead to an empty ``sample_mask`` (:gh:`4558` by `Victoria Shevchenko`_).
```

Comment (Remi-Gau) [reply doc/changes/latest.rst:46]: I double checked and extract_outlier_regressors is not in the public API of nilearn and that's why sphinx cannot find a reference to it and is throwing an error in the doc build. Not mentionning the exact function should fix it.

File: nilearn/interfaces/fmriprep/load_confounds_scrub.py [hunk 4]
```diff
@@ -87,6 +89,15 @@ def extract_outlier_regressors(confounds):
...
     confounds = confounds.loc[:, confounds_cols]
     sample_mask = _outlier_to_sample_mask(outliers)
+
+    if sample_mask is not None and sample_mask.size == 0:
+        warnings.warn(
+            category=RuntimeWarning,
+            message="All time points in the confounds were marked as "
+            "outliers. This would lead to all volumes in the time "
+            "series to be scrubbed.",
+            stacklevel=2,
```
Comment (Remi-Gau) [inline nilearn/interfaces/fmriprep/load_confounds_scrub.py:99]:
```suggestion
            stacklevel=4,
```

Given that extract_outlier_regressors is not part of the public API, users are unlikely to use it directly. I suspect we can raise the stacklevel higher: from a quick check in the codebase I think 4 should do it.

**Comment by victoris93** [comment]
> One easy thing to add yourself as a contributor: add your details in the citation.cff:
> 
> https://github.com/nilearn/nilearn/blob/263ec16553757ef1ba18f76b003b16af264e9fbd/CITATION.cff#L717

Oh I thought it was only for "principal" contributors who author functions and whatnot. Will do.

File: nilearn/interfaces/fmriprep/tests/test_load_confounds_strategy.py [hunk 5]
```diff
@@ -47,6 +47,7 @@ def test_load_confounds_strategy(
         ("ica_aroma", "ica_aroma"),
     ],
 )
+@pytest.mark.filterwarnings("ignore::RuntimeWarning")
```
[the remark this answers is not in the document]
Comment (victoris93) [reply nilearn/interfaces/fmriprep/tests/test_load_confounds_strategy.py:50]: Yes I thought that too.

**Comment by victoris93** [comment]
@Remi-Gau I think should be good. Let me know if I need to do anything else.

**Review by Remi-Gau — APPROVED** [review]
Yup it looks good to me. Thanks for reporting and helping with this.

File: nilearn/signal.py [hunk 6]
```diff
@@ -829,6 +830,8 @@ def _handle_scrubbed_volumes(
     """Interpolate or censor scrubbed volumes."""
     if sample_mask is None:
         return signals, confounds, sample_mask
+    elif sample_mask.size == 0:
```
Comment (bthirion) [inline nilearn/signal.py:833]: My understanding is sample_mask can be an array with 0's, in which case its size is nonzero even if all volumes are removed ?

**Review by bthirion — COMMENTED** [review]
LGTM besides a small detail concerning the different ways to specify sample_mask.

File: nilearn/signal.py (hunk 6 above)
Comment (victoris93) [reply nilearn/signal.py:833]: [Here](https://github.com/nilearn/nilearn/blob/main/nilearn/interfaces/fmriprep/load_confounds_scrub.py#L129) it looks like sample_mask contains the indices of the volumes which are *not* outliers. So the size is only nonzero if at least one volume is not an outlier. Does it make sense?

File: nilearn/_utils/exceptions.py [hunk 7]
```diff
@@ -66,3 +66,20 @@ def message(self):
...
+    """Warns the user if no volumes were kept.
+
+    Exception is raised when all volumes are scrubbed, i.e.,
+    `sample_mask` is an empty array.
+    """
+
+    def __init__(self):
+        super().__init__(
+            "All volumes were scrubbed, can not proceed. "
+            "The size of the sample mask is 0."
```
Comment (htwangtw) [inline nilearn/_utils/exceptions.py:81]: I would prefer the wording here to be more consistent with the below run time warning, and state what is happening with the data first, and then give the implication:
```suggestion
            "The size of the sample mask is 0. "
            "All volumes were marked as motion outliers, can not proceed. "
```

File: nilearn/interfaces/fmriprep/load_confounds_scrub.py [hunk 8]
```diff
@@ -87,6 +89,15 @@ def extract_outlier_regressors(confounds):
         outliers = pd.DataFrame()
     confounds = confounds.loc[:, confounds_cols]
     sample_mask = _outlier_to_sample_mask(outliers)
+
+    if sample_mask is not None and sample_mask.size == 0:
+        warnings.warn(
+            category=RuntimeWarning,
+            message="All time points in the confounds were marked as "
+            "outliers. This would lead to all volumes in the time "
+            "series to be scrubbed.",
```
Comment (htwangtw) [inline nilearn/interfaces/fmriprep/load_confounds_scrub.py:98]: I would keep only volume or time points, rather than using them interchangeably.
Also we are technically marking the volume as the motion outliers, not the confounds.

```suggestion
            message="All volumes were marked as motion outliers. "
            "This would lead to all volumes in the time "
            "series to be scrubbed.",

```

File: nilearn/interfaces/fmriprep/tests/test_load_confounds_scrub.py [hunk 9]
```diff
@@ -96,3 +96,32 @@ def test_extract_outlier_regressors(rng):
...
+    scrub_vol = pd.DataFrame(
+        np.eye(n_scans)[:, idx_scrubbed],
+        columns=[f"{outlier_type}{i:02d}" for i in range(len(idx_scrubbed))],
+    )
+
+    srub_conf = pd.concat([fake_confounds, scrub_vol], axis=1)
+
+    with pytest.warns(
+        RuntimeWarning,
+        match="All time points in the confounds were marked as outliers",
```
Comment (htwangtw) [inline nilearn/interfaces/fmriprep/tests/test_load_confounds_scrub.py:124]: Update based on the suggestions above

```suggestion
        match="All volumes were marked as motion outliers",
```

File: nilearn/signal.py (hunk 6 above)
Comment (htwangtw) [reply nilearn/signal.py:833]: @victoris93 is correct here with how this is `sample_mask` implemented: it's either `None`, meaning no scrubbing, or index of volumes kept.
However this does seem to confuse people quite a lot....

**Review by htwangtw — COMMENTED** [review]
I have some suggestions to the wording consistency with the warning/exception messages

File: nilearn/signal.py (hunk 6 above)
Comment (bthirion) [reply nilearn/signal.py:833]: OK, I understood that the parameter has been coerced to this form when it reaches this function. It is hard to get this from the code.

**Comment by victoris93** [comment]
@htwangtw thanks for the feedback. I think should be good now.

**Comment by Remi-Gau** [comment]
OK CI is green and all review concern have been adressed: merging.
Thanks @bthirion and @htwangtw for the reviews.
Thanks @victoris93 for the PR
